ccp/src/hashrate/hashrate_collector.rs

```diff
@@ -0,0 +1,240 @@
+                    }
+                    ParameterStatus::NotMeasured => ParameterStatus::NotMeasured,
+                };
+
+                let dataset_initialization = match info.dataset_initialization.iter().max() {
+                    Some(&duration) => ParameterStatus::Measured(duration),
+                    None => ParameterStatus::NotMeasured,
+                };
+
+                let statistics = CUHashrate {
+                    effective_hashrate: info.hashes_checked as f64 / epoch_duration,
+                    hashrate,
+                    cache_creation: info.cache_creation,
+                    dataset_initialization,
+                };
+
+                (core_id, statistics)
+            })
+            .collect::<HashMap<_, _>>()
+    }
+
+    pub(crate) fn proof_found(&mut self, core_id: LogicalCoreId) {
+        use std::collections::hash_map::Entry;
+
+        match self.entries.entry(core_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(CUHashrateRaw::from_proof_found());
+            }
+            Entry::Occupied(entry) => entry.into_mut().count_proof_found(),
+        };
+    }
+
+    fn new_epoch(&mut self, epoch: EpochParameters) {
+        self.status = CollectorStatus::Started {
+            started_time: Instant::now(),
+            epoch,
+        };
+
+        self.entries.clear();
+    }
+}
+
+impl Default for CUHashrateRaw {
+    fn default() -> Self {
+        Self {
+            cache_creation: ParameterStatus::NotMeasured,
+            dataset_initialization: vec![],
+            pow_duration: ParameterStatus::NotMeasured,
+            hashes_checked: 0,
+            proofs_found: 0,
+        }
+    }
+}
+
+impl CUHashrateRaw {
+    pub(self) fn from_single_record(entry: HashrateCURecord) -> Self {
+        match entry.variant {
+            HashrateCUEntryVariant::CacheCreation => Self::from_cache_creation(entry.duration),
+            HashrateCUEntryVariant::DatasetInitialization { .. } => {
+                Self::from_dataset_initialization(entry.duration)
+            }
+            HashrateCUEntryVariant::HashesChecked { hashes_count } => {
+                Self::from_cc_job(entry.duration, hashes_count as u64)
+            }
+        }
+    }
+
+    pub(self) fn count_record(&mut self, new_entry: HashrateCURecord) {
+        match new_entry.variant {
+            HashrateCUEntryVariant::CacheCreation => {
+                self.cache_creation = ParameterStatus::Measured(new_entry.duration)
+            }
+            HashrateCUEntryVariant::DatasetInitialization { .. } => {
+                self.dataset_initialization.push(new_entry.duration)
+            }
+            HashrateCUEntryVariant::HashesChecked { hashes_count } => {
+                let new_duration = match self.pow_duration {
+                    ParameterStatus::Measured(duration) => duration + new_entry.duration,
+                    ParameterStatus::NotMeasured => new_entry.duration,
+                };
```

Comment:
Perhaps, we should just use `Option`.

Reply:
I don't like `Option` here b/c their variants are too general and doesn't represent the specific.